vault/request_handling.go

```diff
@@ -1687,10 +1687,11 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
 		// Attach the display name, might be used by audit backends
 		req.DisplayName = auth.DisplayName
 
-		// If this is not a role-based quota, we still need to associate the
-		// login role with this lease for later lease-count quotas to be
-		// accurate.
-		if reqRole == nil && resp.Auth.TokenType != logical.TokenTypeBatch {
+		// If this is not a role-based quota,
+		// we still need to associate the login role with this lease for later
+		// lease-count quotas to be accurate. Unless disable lease role tracking is enabled,
+		// then skip calculating role.
+		if !c.impreciseLeaseRoleTracking && reqRole == nil && resp.Auth.TokenType != logical.TokenTypeBatch {
```

Comment:
We should still take this branch if there is a lease-count quota that's role based, are we missing an extra condition here, or am I misunderstanding the implications of `reqRole == nil`?

Reply:
`reqRole == nil` implies that no role-based quota currently exists on this mount, and thus we haven't looked up the role ahead of time.

We might want to  update the comment to reflect that.

Reply:
Note that if the request mountpath has a role-based RLQ, we'll still be tracking the login role in the `leaseEntry` (since we already did the ResolveRoleOperation).

Reply:
Thanks for calling this out. I switched the ordering of the logic and think its easier to understand now. If `reqRole` is nil it means that it there are no role based quotas (specifically no role based RLQ), so the role has not been added to the request context yet.  If `c.impreciseLeaseRoleTracking` is true, then we can exit because we don't want to keep track of a lease's role. Alternatively, if `c.impreciseLeaseRoleTracking` is false AND the request is a login request, we should calculate the role and add it to context. 

Let me know if that makes any more sense. It was helpful for me to write it out :)

Reply:
I added this comment here too which should emphasize what reqRole implies: https://github.com/hashicorp/vault/pull/22651/files#diff-278ba6b34b839340763b2dc9d4278acf4d4b5826d5d41b13dc81d584e5d446bcR71